replace react-motion with react-spring

drcmda opened the pull request:
This PR removes react-motion and replaces it with react-spring (see: https://github.com/atlassian/react-beautiful-dnd/issues/437).

Since the api is almost the same it was pretty straight forward. I removed react-motion specific work arounds as well.

In a later stage we can think about how to enable native rendering.

drcmda commented:
Also, `translate` should probably be translate3d so we get the gpu to kick in? Or was that a deliberate choice?

TrySound commented:
Test will ask you to update size snapshot. You can do this by running `yarn build` script.

drcmda commented:
As for the tests flagged red, now i don't know if these test against the react-motion workaround i saw in the code or if it's something else. I'd be glad to receive help here. 

@TrySound I suggest once you're happy with your optimizations we could make a build, so we get the benefits of a small, tree-shaken build.

drcmda commented:
Fixed a bunch of complaints. Scratching my head over the remaining few, where it expects it to sit at 0/0 but it's at 75/75 - was that the react-motion bug?

Also added precision to decrease the time it takes for the element to settle. Can be tweaked some more.

alexreardon commented on .size-snapshot.json:
```diff
@@ -1,21 +1,21 @@
-    "bundled": 176774,
-    "minified": 88946,
-    "gzipped": 22589,
+    "bundled": 176172,
```
moving to react-spring would actually result in a minor drop of kb's 👍 

not for the UMD build but i am not too worried about that one

TrySound replied:
I already achieved better results in master. UMD build is bigger because of babel runtime helpers which are not reused.

drcmda commented:
Some interesting development in here: https://github.com/drcmda/react-spring/pull/48

Still i have no idea how to delegate `native` to the end-user, but if there's a solution one day, then with that pr not even mousemove will cause rendering, the Spring would basically bind a `Value` object to a view, and steers it outside of the render loop and mouse/touch events.

alexreardon commented:
I am keen to see this one progress. It is looking really promising

alexreardon commented:
Given we can get rid of a lot of hacks without any KB hit it seems like a good move

TrySound commented:
@drcmda Update dependencies please. Your size results are not relevant.

drcmda commented:
@TrySound @alexreardon Found a way to isolate the spring, even remove it from interpolation. The result is ~3kb, basically it's just the math for the spring now. No element names, color interpolation, etc. NumericalSpring is now part of addons, good for cases like this where we only need to animate raw numerical data. It behaves like Spring otherwise with support for configs, native and so on.

Theoretically it can be reduced even further, if onRest/tracking/onFrame isn't needed.

alexreardon commented:
Whoa
On Wed, 18 Apr 2018 at 6:35 pm, Paul Henschel <notifications@github.com>
wrote:

> @TrySound <https://github.com/TrySound> @alexreardon
> <https://github.com/alexreardon> Found a way to isolate the spring, even
> remove it from interpolation. The result is ~3kb, basically it's just the
> math for the spring now. No element names, color interpolation, etc.
> NumericalSpring is now part of addons, good for cases like this where we
> only need to animate raw numerical data. It behaves like Spring otherwise
> with support for configs, native and so on.
>
> Theoretically it can be reduced even further, if onRest/tracking/onFrame
> isn't needed.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/atlassian/react-beautiful-dnd/pull/442#issuecomment-382309295>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACFN7ZuDtUp2ESv74iC_DS48N83juydGks5tpvq-gaJpZM4TWyQD>
> .
>

TrySound commented:
@drcmda Can you rebase and update dependencies, so we could get real size difference.

alexreardon commented:
I am having a look at how the library feels now

alexreardon commented:
I had an initial play with this and things look positive. I actually noticed some minor performance regressions, but i am not sure what that is about. 

This feels like a good move forward as it:

- Removes a lot of hacky react-motion work arounds
- Has an expressive component api
- Seems to work quite nicely with the exisiting setup
- moves to a more maintained project

Things we still need to do:

- I noticed some bugs when dragging cards on examples that had multiple columns
- Further investigate performance 
- Update tests
- Remove all react-motion hacks (i think there is still one in the action creators)

I am keen to see somebody take this up. Unfortunately I am working on a few other things (get excited) so I do not have capacity right now to own this
